src/storage-client/src/client.rs

```diff
@@ -354,6 +354,65 @@ impl PackableStats for SinkStatisticsUpdate {
     }
 }
 
+#[derive(Copy, Clone, Debug, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
+pub enum Status {
+    Starting,
+    Running,
+    Paused,
+    Stalled,
+    Ceased,
+    Dropped,
+}
+
+impl Status {
+    pub fn to_str(&self) -> &'static str {
+        match self {
+            Status::Starting => "starting",
+            Status::Running => "running",
+            Status::Paused => "paused",
+            Status::Stalled => "stalled",
+            Status::Ceased => "ceased",
+            Status::Dropped => "dropped",
+        }
+    }
+
+    /// Determines if a new status should be produced in context of a previous
+    /// status.
+    pub fn superseded_by(self, new: Status) -> bool {
+        match (self, new) {
+            (Status::Dropped, _) => false,
+            (_, Status::Dropped) => true,
+            (Status::Ceased, _) => false,
+            (_, Status::Ceased) => true,
+            // TODO(guswynn): Ideally only `failed` sources should not be marked as paused.
+            // Additionally, dropping a replica and then restarting environmentd will
+            // fail this check. This will all be resolved in:
+            // https://github.com/MaterializeInc/materialize/pull/23013
+            (Status::Stalled, Status::Paused) => false,
+            // Don't re-mark that object as paused.
+            (Status::Paused, Status::Paused) => false,
+            // De-duplication of other statuses is currently managed by the
+            // `health_operator`.
+            _ => true,
+        }
+    }
+}
+
+impl std::str::FromStr for Status {
+    type Err = anyhow::Error;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(match s {
+            "starting" => Status::Starting,
+            "running" => Status::Running,
+            "paused" => Status::Paused,
+            "stalled" => Status::Stalled,
+            "ceased" => Status::Ceased,
+            "dropped" => Status::Dropped,
+            s => return Err(anyhow::anyhow!("{} is not a valid status", s)),
+        })
+    }
+}
```

Comment:
we should but this next to the `to_str` impl, and we should ensure there are docs that explain that they must be maintained in a backwards compatible way

Reply:
Alternatively we can add the proto field number as a column to the status history tables

Reply:
yeah not worth the macro!